apps/ios/SimpleX NSE/NotificationService.swift

```diff
@@ -219,51 +266,82 @@ class NotificationService: UNNotificationServiceExtension {
+            expectedMessages[receiveEntityId] = ExpectedMessage(
+                ntfMessage: ntfMessage,
+                receiveConnId: connEntity.conn.agentConnId,
+                expectedMsgId: expectedMsgId,
+                allowedGetNextAttempts: 3,
+                msgBestAttempt: (
+                    ntfMessage.ntfsEnabled
+                    ? .nseData(.connectionEvent(ntfMessage.user, connEntity))
+                    : .empty
+                ),
+                ready: false
+            )
+        }
+    }
+
     override func serviceExtensionTimeWillExpire() {
         logger.debug("DEBUGGING: NotificationService.serviceExtensionTimeWillExpire")
         deliverBestAttemptNtf(urgent: true)
     }
 
-    func processReceivedNtf(_ ntf: NSENotification) -> Bool {
-        guard let ntfInfo = notificationInfo, let expectedMsgTs = ntfInfo.expectedMsg_?.msgTs else { return false }
+    func processReceivedNtf(_ id: ChatId, _ ntf: NSENotification) -> Bool {
+        guard
+            let expectedMessage = expectedMessages[id],
+            let expectedMsgTs = expectedMessage.ntfMessage.expectedMsg_?.msgTs
+        else {
+            return !expectedMessages.allSatisfy { $0.value.ready }
+        }
+        let ntfInfo = expectedMessage.ntfMessage
         if !ntfInfo.user.showNotifications {
-            self.setBestAttemptNtf(.empty)
+            expectedMessages[id]?.msgBestAttempt = .empty
         }
         if case let .msgInfo(info) = ntf {
-            if info.msgId == expectedMessage {
-                expectedMessage = nil
-                logger.debug("NotificationService processNtf: msgInfo msgId = \(info.msgId, privacy: .private): expected")
+            if info.msgId == expectedMessage.expectedMsgId {
+                expectedMessages[id]?.expectedMsgId = nil
+                expectedMessages[id]?.ready = true
+                logger.debug("NotificationService processNtf: msgInfo msgId = \(info.msgId, privacy: .public): expected")
                 self.deliverBestAttemptNtf()
                 return true
             } else if let msgTs = info.msgTs_, msgTs > expectedMsgTs {
-                logger.debug("NotificationService processNtf: msgInfo msgId = \(info.msgId, privacy: .private): unexpected msgInfo, let other instance to process it, stopping this one")
+                logger.debug("NotificationService processNtf: msgInfo msgId = \(info.msgId, privacy: .public): unexpected msgInfo, let other instance to process it, stopping this one")
+                expectedMessages[id]?.ready = true
                 self.deliverBestAttemptNtf()
-                return false
-            } else if allowedGetNextAttempts > 0, let receiveConnId = receiveConnId {
-                logger.debug("NotificationService processNtf: msgInfo msgId = \(info.msgId, privacy: .private): unexpected msgInfo, get next message")
-                allowedGetNextAttempts -= 1
+                return !expectedMessages.allSatisfy { $0.value.ready }
```

Comment:
Currently just returning `!expectedMessages.allSatisfy { $0.value.ready }`, which means message loop continues while there are not ready "expected message" entries, and that messages previously returning false do not get processed by other threads if they exist. Simply returning false is not correct because then if there are no other matching threads loop will be stuck on waiting for a single message, and batch of expected messages would not finish processing (or would take longer than necessary).

Couple of options:
- We could change return to some enum instead of bool signalling "go onto next", "limited retry" (e.g. when not all expected messages are ready - try once or a few more times to find another thread, but shorter than current wait limit), "wait" (when all expected messages are ready - ok to wait as before).
- We could put these rejected messages to some other queue where they would be waiting for matching threads as before, without holding processing of this batch of expected messages.